WIP: Flow server, missing .flowconfig, auto-install

Adezandee opened the pull request:
Hi there,

I did some rework of this module because i find nuclide too heavy and complex :)
This PR address issues #24, #27.

Additional changes:
* Flow server spawning and lifecycle per project (with notifications)
* Removed babel compilation in favor of atom built-in `'use babel'`
* Renamed config to more standard `'executablePath'`(atom-linter)
* Added some coding convention with .eslintrc and .editorconfig files

@steelbrain if you have any feedback on these changes, i'll be happy to improve this :+1:

steelbrain commented on index.js:
```diff
@@ -0,0 +1,116 @@
+'use babel';
+
+import { CompositeDisposable } from 'atom';
+import { spawn } from 'child_process';
```
I would suggest you do [`atom-linter::exec`](https://npmjs.org/package/atom-linter)

Adezandee replied:
exec won't return to me the node::childprocess object unfortunatly

steelbrain commented:
Looks pretty nice. but can you remove those `console.log`s. Also you don't **have** to spawn the flow server, it's automatically spawned when you do the first lint request. So here's the idea: Try to spawn the server, if you can't parse the response, show all of it as a notification otherwise parse them

Adezandee commented:
Thanks. I know that the server is automatically started with first request but its bugs me that it's still lives even when you close Atom (that the current behavior). This means that you could end up with a bunch of flow server running on your comp ... That's why I wanted to handle serverProcess child lifecycle

Sure I can remove those console.log, since I've seen a bunch of those in other repositories (linter-eslint), I assumed that it was ok

steelbrain commented:
We are slowly removing all of the `console.log`s in linters or at least showing them when atom is in dev mode. About the flow server, I don't think it consumes much resources, besides if it was running before the user started atom, it'll be killed when the user quits atom which might not be what we want. Flow servers are usually super-lightweight so we barely feel them. and they can be used from cli and other flow integrations. so I would advise that we don't kill them :)

Adezandee commented:
Of course, the current implementation can't kill already running flow server. It's really just killing those booted up by the linter himself (if necessary, I'll check how this behaves with a flow server already running in a terminal outside of Atom)

steelbrain commented:
I would advise you do, and report back the results

Adezandee commented:
@steelbrain I updated the PR adding:
* Handle externally running flow server
* Add some flow types to the repository himself

steelbrain commented on index.js:
```diff
@@ -0,0 +1,72 @@
+
+class LinterFlow {
+  config = {
+    pathToFlow: {
```
In all the other linter providers, we have named this config `executablePath`, I would suggest you use that too for consistancy

steelbrain commented on index.js:
```diff
@@ -0,0 +1,72 @@
+      lintOnFly: true,
+      lint: (TextEditor) => {
+        const filePath = TextEditor.getPath();
+        const fileText = TextEditor.buffer && TextEditor.buffer.cachedText;
```
There is **always** a buffer with an editor, so you can just drop the check

steelbrain commented on lib/message.js:
```diff
@@ -0,0 +1,46 @@
+    return [];
+  }
+  return data.errors
+    .map(item => item.message)
```
Instead of doing two iterations, can you just handle `item.message` in the `flowMessageToLinterMessage` function?

steelbrain commented:
We are almost ready to :ship: but I still think that spawning the flow server and then killing it isn't worth it. I am a hack user and I honestly don't mind an `hh_server`, because when I invoke it once, it means I am gonna work and when I am gonna work I am gonna need it again and again. Because the overhead of it is 80 MB RAM (on a quite big project) I think it should be left running, there's no harm in that.

Adezandee commented:
Well, i've been following [nuclide](https://github.com/facebook/nuclide/blob/master/pkg%2Fnuclide%2Fflow-base%2Flib%2FFlowExecutor.js#L24) way of handling this and I think it's better to handle flow server if necessary in case the end-user isn't familiar with it. Every other linter is kinda "install and forget", meaning that if you close Atom, there is no other side-effect.

steelbrain commented:
I am not gonna push too hard on that, I mean, I got no problems either way, I had a bit of a problem with this package terminating external servers but you seem to have taken care of that. The implementation looks nice, now we just need somebody to test it. Mind testing it for your brother fellaws? @Arcanemagus @SpainTrain @kepler0

steelbrain commented on lib/message.js:
```diff
@@ -0,0 +1,44 @@
+
+  return {
+    type: first.level,
+    text: messages.map(msg => msg.descr).join('. '),
```
Mind explaining why you're showing the description from traces in the main message?

Adezandee replied:
It's because the first message description is often only one word and doesn't help understand the problem. concatenating message from traces create a understandable error message (to show in linter bottom pane)

steelbrain replied:
I would love to see an example :)

Adezandee replied:
i'll add some unit test / spec :+1:

Adezandee replied:
@steelbrain nuclide do the same to render message in Panes, [see here](https://github.com/facebook/nuclide/blob/746090718fe19222a88c1d34e23c06428dc294c9/pkg%2Fnuclide%2Fdiagnostics%2Fui%2Flib%2FDiagnosticsPane.js#L75)

nmn commented:
It all looks pretty nice. I probably want to take some of this code over to autocomplete-flow as well.

The flow server but is a little tricky. Maybe that can just be made an option??

Adezandee commented:
Hi @steelbrain, @nmn !

Some new changes:
* Added some units tests extracted from [@flow](https://github.com/facebook/flow)
* Added a new parameters `enableAll`, this adresses issue #25

steelbrain commented:
@Adezandee Nuclide does not support/use traces as far as I know so we shouldn't follow/adopt everything of theirs.

Adezandee commented:
@steelbrain they do support traces in their own custom DiagnosticMessage. But anyway it currently render quite nice, and i'm happy with the overall quality
![screen shot 2015-10-27 at 23 58 55](https://cloud.githubusercontent.com/assets/4415204/10775246/1338c876-7d07-11e5-9367-b18ca2e4a27c.png)

steelbrain commented:
That flow message is kinda funny, I use the hack typechecker which shares the same codebase and it's messages are a **lot** friendlier than that,

Adezandee commented:
I too found flow messages quite cryptic and not very friendly for rendering, here you can see that concatenation helps a bit

steelbrain commented:
I do agree that they should be concatenated, btw, has anyone opened an issue about this on flow repo? The error above should be `It's a number [location]`, `This type is incompatible with string [location]`.

nmn commented:
@Adezandee @steelbrain Nuclide and linter-flow both have the same error parsing for traces.
The messages are pretty unfriendly at the moment, but this is a problem with flow itself.

I say we leave the errors and traces the way they are for now, and create an issue on flow to get it fixed there.

The way the errors work is that the message is broken up over the error and the traces. I concat the messages for the error so it can read like a sentence, but the traces still have cryptic messages. I see no simple way to fix it. At the same time, remove traces is not something we should be doing. Traces are extremely useful when actually developing with flow turned on.

Adezandee commented:
@steelbrain @nmn 
I think that this PR is ready for merge now, i've been testing it this week and it works quite nicely.
Btw, i'm open to take responsibility on this repository if you feel like it, i'll love to enable Travis and improve it further.
Thanks

steelbrain commented:
This PR LGTM. r? @nmn

nmn commented:
I'll give it a quick test and merge it tomorrow.

Adezandee commented:
ping @nmn

nmn commented:
On it.

nmn commented:
@Adezandee some notes:
1) please add `eslint`, `babel-eslint`, `eslint-preset-airbnb` and `eslint-plugin-babel` as a devDependencies.

... will edit to add more comments.

nmn commented on lib/index.js:
```diff
@@ -0,0 +1,84 @@
+        if (!flowConfig) {
+          atom.notifications.addWarning('[Linter-Flow] Missing .flowconfig file.');
+          return [];
+        }
```
There should an config option for linter-flow that disables the linter if no .flowconfig file is present.

nmn commented on lib/index.js:
```diff
@@ -0,0 +1,84 @@
+        const fileText = TextEditor.buffer.cachedText;
+
+        // Is flow enabled for current file ?
+        if (fileText.indexOf('@flow') === -1 && !this.enableAll) {
```
Can we also add a config option to turn this check on or off.

nmn commented:
I really like a lot of the work you have put in. But losing listing on the fly is a deal breaker. Let's get that fixed and reopen this.

Adezandee commented:
@nmn Well I don't get it. I don't see any regression, it is still linting on fly. have you tried it ? btw `check-content` only gives hint about the file considered alone (aka. without project context), not really helpful.

As for the two others points:
* eslint deps: well its not up to us to install eventual flow addons to make it work with .flowconfig. My point  is that if eslint was a python program it would be required to be installed on dev machine by others means. This whole repository assumes that you run on unix env, why not assume you have also eslint and deps installed globally ?
* Option to disable linter if no .flowconfig is present: yes of course, I could be annoying to see the warning message constantly pop-in.

Adezandee commented:
Also, I find the "PR closed" action of yours kinda harsh...

steelbrain commented:
@Adezandee I do agree it's a bit rude, he could have just added `WIP` to the title instead of closing it

nmn commented:
@Adezandee sorry if you thought it was harsh. That was not the intention. You've done really great work here. But listing on the fly does not work anymore. You can try your self. Purposely write some code that would be an error.

`var x: string = 1`

Then when you see the error. Try adding quotes around the 1. The error won't go away unless you save. Without flow check-contents there is no way of passing the contents of the unsaved file to flow.

I do agree that it only checks the current file, but I think that's worth it.

As for installing eslint as a Dev dependency, that's a relatively minor issue. It just adds some stuff to the package.json file and makes life easier for contributors.

nmn commented:
Btw, check-content works in context of the entire project. It only returns the errors for the current file but the flow server takes the entire project into consideration.

nmn commented:
Other than the things mentioned already:
- The flow comment should be added `index.js` and `servers.js`
- A declarations file should be added to declare the globals and modules provided by atom.

On my machine I wrote a declarations file that looks like this:
```
/* eslint-disable */

declare var atom: Object;

declare class Subscription {
  dispose(): void;
}

declare class Config {
  observe(config: string, cb: Function): Subscription;
}

declare class CompositeDisposableClass {
  add(observable: Subscription): void;
  dispose(): void;
}


declare module atom {
  declare var Range: any;
  declare var CompositeDisposable: any;
  declare var config: Config;
}

declare module 'atom-package-deps' {
  declare function install(packageName: string): void;
}

declare module 'atom-linter' {
  declare function findFile(filePath: string, fileName: string): ?string;
  declare function exec(executable: string, args?: Array<string>, config?: Object): Promise<string>;
}
```

And the `.flowconfig` file now looks like this:

```
/* eslint-disable */

declare var atom: Object;

declare class Subscription {
  dispose(): void;
}

declare class Config {
  observe(config: string, cb: Function): Subscription;
}

declare class CompositeDisposableClass {
  add(observable: Subscription): void;
  dispose(): void;
}


declare module atom {
  declare var Range: any;
  declare var CompositeDisposable: any;
  declare var config: Config;
}

declare module 'atom-package-deps' {
  declare function install(packageName: string): void;
}

declare module 'atom-linter' {
  declare function findFile(filePath: string, fileName: string): ?string;
  declare function exec(executable: string, args?: Array<string>, config?: Object): Promise<string>;
}
```

I've also added a few type annotations in the files that are now type checked.

Of course the declarations can be improved, but this was a good start.

nmn commented:
@steelbrain @Adezandee Can we maybe create a new branch with the changes in this PR, so we can collaborate and fix the things are left over. Pasting files here feels stupid.

steelbrain commented:
@nmn I do agree that a branch would be nice. He PRed this when he wasn't in the team so we have two options now

1. He adds you as a collaborator to his repo and you can fix it there and it'll get synced here
2. This PR is closed and a new PR with a branch of this repo is opened with the same diff

nmn commented:
I prefer option 2.

Adezandee commented:
Option 2 is fine to me :+1:

steelbrain commented:
Option 2 it is then :clap:

nmn commented:
Ping me when you add the branch on here @Adezandee. I've already made some fixes locally. It should be quick work.
